crates/ark/src/analysis/parse_boundaries.rs

```diff
@@ -0,0 +1,375 @@
+            return;
+        };
+        incomplete = Some(std::ops::Range { start, end: *end });
+    };
+
+    for current_line in (0..n_lines).rev() {
+        let mut record_complete = |exprs: RObject| -> anyhow::Result<()> {
+            let srcrefs = exprs.srcrefs()?;
+            let mut ranges: Vec<std::ops::Range<usize>> =
+                srcrefs.into_iter().map(|srcref| srcref.line).collect();
+
+            complete.append(&mut ranges);
+            Ok(())
+        };
+
+        // Grab all code up to current line
+        let subset = &lines_r.slice()[..current_line + 1];
+        let subset = CharacterVector::try_from(subset)?;
+
+        // Parse within source file to get source references
+        let srcfile = harp::srcref::SrcFile::try_from(&subset)?;
+
+        match harp::parse_status(&harp::ParseInput::SrcFile(&srcfile))? {
+            ParseResult::Complete(exprs) => {
+                record_complete(exprs)?;
+                record_incomplete(current_line + 1, &incomplete_end);
+                record_error(current_line + 1, &error_end);
+                break;
+            },
+
+            ParseResult::Incomplete => {
+                record_error(current_line + 1, &error_end);
+
+                // Declare incomplete
+                if let None = incomplete_end {
+                    incomplete_end = Some(current_line + 1);
+                }
+            },
+
+            ParseResult::SyntaxError { .. } => {
+                // Declare error
+                if let None = error_end {
+                    error_end = Some(current_line + 1);
+                }
+            },
+        };
+    }
+
+    // Note that these are necessarily no-ops if we exited the loop after
+    // detected complete expressions since we already recorded the boundaries in
+    // that case.
+    record_incomplete(0, &incomplete_end);
+    record_error(0, &error_end);
+
+    // Merge expressions separated by semicolons
+    let complete = merge_overlapping(complete);
+
+    let boundaries = ParseBoundaries {
+        complete,
+        incomplete,
+        error,
+    };
+
+    // Fill any gaps with one-liner complete expressions
+    let boundaries = fill_gaps(boundaries, n_lines);
+
+    Ok(boundaries)
+}
+
+fn merge_overlapping<T>(ranges: Vec<std::ops::Range<T>>) -> Vec<std::ops::Range<T>>
+where
+    T: PartialOrd + Ord + Copy,
+{
+    let merge = |mut merged: Vec<std::ops::Range<T>>, range: std::ops::Range<T>| {
+        if let Some(last) = merged.last_mut() {
+            // if range.start <= last.end {
+            if last.contains(&range.start) {
+                // Overlap, merge with last range
+                last.end = last.end.max(range.end);
+            } else {
+                // No overlap, push a new range
+                merged.push(range);
+            }
+        } else {
+            // First element, just add it
+            merged.push(range);
+        }
+
+        merged
```

Comment:
```suggestion
        let Some(last) = merged.last_mut() else {
            // First element, just add it
            merged.push(range);
            return merged;
        };

        // if range.start <= last.end {
        if last.contains(&range.start) {
            // Overlap, merge with last range
            last.end = last.end.max(range.end);
        } else {
            // No overlap, push a new range
            merged.push(range);
        }

        merged
```

I felt like it was slightly easier to understand if you get the early exit out of the way